src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,359 @@
+				),
+				SyntaxKind.InvocationExpression
+			);
+
+			// After the rest of analysis report about potentially unnecessary [Blocking]s
+			ctx.RegisterCompilationEndAction(
+				ctx => WarnForUnusedBlockingAttributes(
+					ctx,
+					attr,
+					havePossiblyUnusedBlockingAttribute: havePossiblyUnusedBlockingAttribute.Keys,
+					callsSomethingBlocking: callsSomethingBlocking.Keys
+				)
+			);
+		}
+
+		private static void InspectMethodDeclaration(
+			SymbolAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodWithPossiblyUnusedBlockingAttribute
+		) {
+			var methodSymbol = (IMethodSymbol)ctx.Symbol;
+
+			// For non-async methods, look at the things we implement and
+			// whether they are blocking or not.
+			var implementedThings = methodSymbol.GetImplementedMethods();
+
+			var implementedBlockingThings = new List<IMethodSymbol>();
+			var implementedNonBlockingThings = new List<IMethodSymbol>();
+
+			foreach( var implementedMethod in implementedThings ) {
+				if( HasAttribute( implementedMethod, blockingAttr ) ) {
+					implementedBlockingThings.Add( implementedMethod );
+				} else {
+					implementedNonBlockingThings.Add( implementedMethod );
+				}
+			}
+
+			var attrData = GetAttribute( methodSymbol, blockingAttr );
+
+			if( attrData == null ) {
+				if( implementedBlockingThings.Count == 0 ) {
+					// We don't have [Blocking] and none of the things we implement are
+					// blocking, so there is nothing to do here.
+					return;
+				} else {
+					// We implement something blocking but don't have [Blocking]. That's
+					// fine: direct callers get to know that we're not actually blocking,
+					// and indirect callers have to assume the worse.
+					// Still output a diagnostic so that we can suggest a code fix in the
+					// IDE to add [Blocking].
+
+					// TODO: code fix to add [Blocking]
+
+					var decl = methodSymbol.DeclaringSyntaxReferences.First().GetSyntax() as MethodDeclarationSyntax;
+
+					ctx.ReportDiagnostic(
+						Diagnostic.Create(
+							Diagnostics.NonBlockingImplementationOfBlockingThing,
+							decl.Identifier.GetLocation(),
+							$"{methodSymbol.ContainingType.Name}.{methodSymbol.Name}",
+							$"{implementedBlockingThings[0].ContainingType.Name}.{implementedBlockingThings[0].Name}"
+						)
+					);
+					return;
+				}
+			}
+
+			// So we do have a [Blocking] attribute.
+
+			if( implementedBlockingThings.Count == 0 ) {
+				// We're not forced to have [Blocking] due to what we implement, so check
+				// if we need it at all.
+				noteMethodWithPossiblyUnusedBlockingAttribute( methodSymbol );
+			}
+
+			if( methodSymbol.IsAsync ) {
+				// TODO: code fix to remove [Blocking]
+
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.AsyncMethodCannotBeBlocking,
+						attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+						methodSymbol.Name
+					)
+				);
+				return;
+			}

```

Comment:
Should this more generically be "returns awaitable" rather than relying on the async modifier itself?

Reply:
Similar logic would be needed if we wanted to make this an error:

```csharp
public IFoo {
  [Blocking]
  Task BarAsync();
}
```

Reply:
I had thought about it and decided not to. If we did it would probably be better to do `IAsyncResult` instead, but admittedly the difference isn't going to matter for D2L. I decided not to because it's not immediately trouble to do blocking calls and return a `Task`. It's very likely the case that if you were doing this you'd end up `await`ing anyway (and get the error) but if not, I mean, maybe you're fine? You can certainly imagine scenarios where its fine, but I don't think they are relevant for D2L either way.

Reply:
So if we never awaited something in the whole stack not encountering the error, is it problematic to return a "blocking Task" in an async controller?

Either way, you're right it should get "caught eventually", happy to defer.

Reply:
Possibly not as bad, but it really depends. I switched it to `IAsyncResult` return vals.